source/extensions/filters/http/common/compressor/compressor.h

```diff
@@ -0,0 +1,197 @@
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Common {
+namespace Compressors {
+
+/**
+ * All compressor filter stats. @see stats_macros.h
+ * "total_uncompressed_bytes" only includes bytes from requests that were marked for compression.
+ * If the request was not marked for compression, the filter increments "not_compressed", but does
+ * not add to "total_uncompressed_bytes". This way, the user can measure the memory performance of
+ * the compression.
+ * "header_gzip" is specific to the gzip filter and is deprecated since it duplicates
+ * "header_compressor_used".
+ */
+// clang-format off
+#define ALL_COMPRESSOR_STATS(COUNTER)    \
+  COUNTER(compressed)                    \
+  COUNTER(not_compressed)                \
+  COUNTER(no_accept_header)              \
+  COUNTER(header_identity)               \
+  COUNTER(header_gzip)                   \
+  COUNTER(header_compressor_used)        \
+  COUNTER(header_compressor_overshadowed)\
+  COUNTER(header_wildcard)               \
+  COUNTER(header_not_valid)              \
+  COUNTER(total_uncompressed_bytes)      \
+  COUNTER(total_compressed_bytes)        \
+  COUNTER(content_length_too_small)      \
+  COUNTER(not_compressed_etag)           \
+// clang-format on
+
+/**
+ * Struct definition for compressor stats. @see stats_macros.h
+ */
+struct CompressorStats {
+  ALL_COMPRESSOR_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+class CompressorFilterConfig {
+
+public:
+  CompressorFilterConfig() = delete;
+
+  virtual const std::string featureName() const PURE;
+  virtual std::unique_ptr<Compressor::Compressor> makeCompressor() PURE;
+
+  Runtime::Loader& runtime() { return runtime_; }
+  CompressorStats& stats() { return stats_; }
+  const StringUtil::CaseUnorderedSet& contentTypeValues() const { return content_type_values_; }
+  bool disableOnEtagHeader() const { return disable_on_etag_header_; }
+  bool removeAcceptEncodingHeader() const { return remove_accept_encoding_header_; }
+  uint32_t minimumLength() const { return content_length_; }
+  const std::string contentEncoding() const { return content_encoding_; };
+  const std::map<std::string, uint32_t> registeredCompressors() const;
+
+  virtual ~CompressorFilterConfig();
+
+protected:
+  CompressorFilterConfig(const envoy::config::filter::http::compressor::v2::Compressor& compressor,
+                         const std::string& stats_prefix, Stats::Scope& scope, Runtime::Loader& runtime,
+                         const std::string& content_encoding);
+
+private:
+  static StringUtil::CaseUnorderedSet
+  contentTypeSet(const Protobuf::RepeatedPtrField<std::string>& types);
+
+  static uint32_t contentLengthUint(Protobuf::uint32 length);
+
+  static CompressorStats generateStats(const std::string& prefix, Stats::Scope& scope) {
+    return CompressorStats{ALL_COMPRESSOR_STATS(POOL_COUNTER_PREFIX(scope, prefix))};
+  }
+
+  struct CompressorRegistry {
+    Thread::MutexBasicLockable mutex_;
```

Comment:
Sorry I should've asked this earlier as I think a lot of my concern flows from this. Can you describe the life of a CompressorFilterConfig?

When do these get discovered? Can we grow a new compressor after requests start to flow? If so, how?

IIUC you as it stands you will acquire this global mutex on every request, which is an anti-pattern in Envoy for performance on machines with lots of CPUs.

Reply:
Oh right -- sorry I meant to ask about CompressorRegistry. I'd really like to see this locked in during binary startup and considered const thereafter. Is that feasible?

Reply:
Probably this could be done with `Init::Manager`. But I've just realized there is a bigger problem: there could be many filter chains and filters need to be aware of its siblings only, therefore the current approach with global static `CompressorRegistry` is wrong and applicable only for the case of a single listener.

Looks like we need to let filter factories know about the chains they are instantiated for. But that's rather an invasive change. Would it make sense?

Reply:
Sorry I should've responded to this earlier that I'm not 100% sure if that change is a good idea globally in Envoy, or needed for this problem.

@mattklein123 for filter architecture issues.

In my thinking, the best time to set up the compressor registry is before any requests start flowing. Once it's set up it should be const. In my ideal world this would be done in MainContextBase's constructor or similar. But I think a lot of auto-registered things happen magically before main() runs due to static constructors, and we can make that work as long as we are hyper-careful about static-init-order fiascos.

Reply:
Sorry this is a big PR. Can you summarize what part I should look at? Thank you.

Reply:
Matt -- no need to look at the PR; just the comment from rojkov about Init::Manager and filter chains, etc. I didn't fully understand the proposal or why it's needed.

Reply:
I think I need more info than what is in this comment thread. @rojkov can you summarize?

Reply:
@mattklein123 This PR generalizes HTTP compression filters into a single class (`CompressionFilter`) which can be parameterized with a compressor object, e.g. `Compressor::ZlibCompressorImpl` or `Compressor::BrotliCompressorImpl`. All is good when there's only one compression filter in a chain.

```yaml
static_resources:
  listeners:
    name: listener_0
    address:
      socket_address:
        address: 127.0.0.1
        port_value: 10000
    filter_chains:
      filters:
      - name: envoy.http_connection_manager
        config:
          codec_type: auto
          http_filters:
          - name: envoy.gzip
            config: {}
          - name: envoy.router
            config: {}
...
```

Now let's consider the case
```yaml
static_resources:
  listeners:
    name: listener_0
    address:
      socket_address:
        address: 127.0.0.1
        port_value: 10000
    filter_chains:
      filters:
      - name: envoy.http_connection_manager
        config:
          codec_type: auto
          http_filters:
          - name: envoy.gzip
            config: {}
          - name: envoy.brotli
            config: {}
          - name: envoy.router
            config: {}
...
```

and a client making a request with
```
Accept-Encoding: br;q=0.75, gzip;q=0.25
```

In this case `envoy.gzip` shouldn't modify the stream and let `envoy.brotli` do compression. For this `envoy.gzip` should be aware of `envoy.brotli` sitting in the chain.

To solve the problem this PR introduces a global single registry (a flat map) `CompressorRegistry` where the compression filters register themselves which turned out to be a bad idea, because there might be more than one listener configured. And every listener may have at least one separate filter chain like in
```yaml
static_resources:
  listeners:
  - name: listener_0
    address:
      socket_address:
        address: 127.0.0.1
        port_value: 10000
    filter_chains:
      filters:
      - name: envoy.http_connection_manager
        config:
          codec_type: auto
          http_filters:
          - name: envoy.gzip
            config: {}
          - name: envoy.brotli
            config: {}
          - name: envoy.router
            config: {}
...
  - name: listener_1
    address:
      socket_address:
        address: 127.0.0.1
        port_value: 11000
    filter_chains:
      filters:
      - name: envoy.http_connection_manager
        config:
          codec_type: auto
          http_filters:
          - name: envoy.brotli
            config: {}
          - name: envoy.router
            config: {}
...
```

`envoy.brotli` from `listener_1` must ignore `envoy.gzip` from `listener_0` when handling requests with
```
Accept-Encoding: br;q=0.2, gzip
```

One solution could be that we let compression filters know which filter chain they belong.

Another solution is to introduce one more level of abstraction: an HTTP filter which is a container for all compression filters in the chain:
```yaml
static_resources:
  listeners:
    name: listener_0
    address:
      socket_address:
        address: 127.0.0.1
        port_value: 10000
    filter_chains:
      filters:
      - name: envoy.http_connection_manager
        config:
          codec_type: auto
          http_filters:
          - name: envoy.http_compression_manager
            config:
              http_compression_filters:
              - name: envoy.compression.gzip
                config: {}
              - name: envoy.compression.brotli
                config: {}
          - name: envoy.router
            config: {}
...
```
Then we don't have to deal with the singleton registry at the expense of more complicated configs.

Reply:
Ah this clears it up for me too.

WDYT of having a single filter manages both gzip & brotli (or *all* compressors more generally if there's ever another alternative) and makes intelligent decisions based on the Accept-Encoding header, about which to apply.

Reply:
Yes, that's basically the second proposal described above.

I hope I can find time to rework it all :)

Reply:
Yup that makes sense. A cheaper solution would also be to piggy back on to: https://github.com/envoyproxy/envoy/pull/8851. This introduces a per-request named state object, which you could use to grab state from each compression filter.

Reply:
Would it be sufficient, with this architecture, to depend on the user to simply put the higher priority filter first in the chain? So if brotli is first, and it runs, it will leave content-encoding:br in the response headers, which could help gzip decide to abstain from further compression.

Reply:
There are two phases: decoding request and encoding response. Filters need to decide which one is going to be used during the first phase (when handling `Accept-Encoding`). In this phase the order of the filters may matter only in case of something like `Accept-Encoding: *`. But in the second phase the order becomes very important and if a filter sees `Content-Encoding` set it just passes the stream through. I have a working prototype for symmetric compression/decompression and it relies on the order of filters in the chain to support use cases like ["decompress the gzipped stream received from `cluster_0` and compress it again with brotli if the http-client accepts it"](https://github.com/envoyproxy/envoy/issues/4445#issuecomment-509640676).

Reply:
@mattklein123 I've looked at #8851 and it is a much cheaper solution indeed. Actually even `Envoy::StreamInfo` would suffice if set at the moment of setting decoder filter callbacks (before stream processing starts). I'll give it a try.